🐛 Fix TypeScript  following #1318

paulrostorp opened the pull request:
This PR aims to revert breaking changes and broken types following #1318

paulrostorp commented on .eslintrc.yml:
```diff
@@ -111,6 +111,11 @@ overrides:
       - 'src/**/*.ts'
     rules:
       flowtype/no-types-missing-file-annotation: off
+  - files:
+      - '**/*d..ts'
+    rules:
+      no-unused-vars: off
```
Eslint was throwing unnecessary errors because of this.

radex commented:
@enahum Can you help with reviewing this PR? Does it look good to you?

enahum commented:
> @enahum Can you help with reviewing this PR? Does it look good to you?

Will do as soon as I have some time available later today

radex commented:
@paulrostorp Take a look at the CI issue - eslint changes are broken?

paulrostorp commented:
@radex My bad, I had a typo [301e814](https://github.com/Nozbe/WatermelonDB/pull/1477/commits/301e814c77ab742bc33999373e27c338196913ab)

radex commented:
@paulrostorp Yes, don't worry about that. But can you take a look at one more thing? The original contributor of TS added this: `yarn test:typescript`, but this never was integrated into the CI pipeline, so I'm not sure it's not broken. If this can be made to work, we should add it to `yarn ci:check` so that TS types can be ensured to be more up to date

paulrostorp commented:
> @paulrostorp Yes, don't worry about that. But can you take a look at one more thing? The original contributor of TS added this: `yarn test:typescript`, but this never was integrated into the CI pipeline, so I'm not sure it's not broken. If this can be made to work, we should add it to `yarn ci:check` so that TS types can be ensured to be more up to date

Ah cool That's actually highlighted quite a few things

enahum commented on src/Model/index.d.ts:
```diff
@@ -52,14 +52,14 @@ export default class Model {
   //   task.name = 'New name'
   // })
-  update(recordUpdater: (this) => void): Promise<this>
+  update(recordUpdater: (_: this) => void): Promise<this>
```
the function used to be optional

enahum commented on src/Model/index.d.ts:
```diff
@@ -52,14 +52,14 @@ export default class Model {
   // After preparing an update, you must execute it synchronously using
   // database.batch()
-  prepareUpdate(recordUpdater: (this) => void): this
+  prepareUpdate(recordUpdater: (_: this) => void): this
```
function should be optional, default is noop

enahum commented on src/Database/index.d.ts:
```diff
@@ -16,7 +16,7 @@ import { $ReadOnlyArray, $Exact } from '../types'
 type DatabaseProps = $Exact<{
   adapter: DatabaseAdapter
-  modelClasses: Model[]
+  modelClasses: typeof Model[]
```
this is causing problems when doing `new Database({adapter, models})` where models is of type `Model[]`

paulrostorp replied:
that's odd... What error do you get ? 

I was getting an error _before_ making this change...

btw it should be `new Database({adapter, modelClasses})`

radex replied:
I don't really know TypeScript, but it surely can't be `Model[]` -- should't it be `class<Model>[]`?

paulrostorp replied:
I've reverted to the original custom `Class` type just to be safe. I suspect there may be something with different typescript versions...

enahum reviewed (commented):
Also one remaining issue is that if I create a type for a model that extends on `Model` and a class that implements such type, then typescript complains with something like

```
Type 'typeof InfoModel' is missing the following properties from type 'Model': _raw, _isEditing, _preparedState, _getChanges, and 30 more.ts(2740)
```

radex commented on .github/workflows/ci.yml:
```diff
@@ -21,6 +21,8 @@ jobs:
       - run: yarn
       - run: yarn ci:check
+      - name: TS Checks
+        run: yarn test:typescript
```
add this to ci:check instead

paulrostorp replied:
done

radex commented on package.json:
```diff
@@ -16,7 +16,7 @@
     "test:ios": "scripts/test-ios",
     "test:native": "concurrently -n android,ios 'npm run test:android' 'npm run test:ios' --kill-others-on-fail",
-    "test:typescript": "cd examples/typescript; yarn; npm run test",
+    "test:typescript": "yarn build; cd examples/typescript; yarn; npm run test",
```
hmm, building is necessary?

paulrostorp replied:
Because I setup the package as a link (https://github.com/Nozbe/WatermelonDB/pull/1477/files#diff-52a1428e31d738b6bcc8a9c9c840fa66484d5bf4e84921e0628eeb7596148371R9), I chose to use the production output as it best represents the final folder stucture.

I could map it to source though, but eslint will complains (although typescript _should_ work)

paulrostorp replied:
I had tried it here: https://github.com/Nozbe/WatermelonDB/pull/1477/commits/b04a356af5907b97cb502ecd2ebc925e99df5637

radex replied:
If possible, I'd rather not run `yarn build` on `yarn ci:check`. Having `dist/` causes issues with other tools :/

paulrostorp replied:
@radex I disabled eslint for the ts example app and made it work using typescript paths in [597e3fe](https://github.com/Nozbe/WatermelonDB/pull/1477/commits/597e3fe2b0c0e408e7a9fe969331d98b3af6ebde)

radex reviewed (approved):
LGTM, tooling-wise. I'm not making any judgements on the TS types themselves, like I said, I don't use them, and don't have TS experience.

radex commented:
I'm OK with merging this, but would be nice to have a second opinion. @enahum Does this version work for you?

enahum commented:
I'll give it a try tomorrow if you guys can hold it until then

radex commented:
@enahum Yes. I won't be releasing today

radex commented:
@enahum Hey, any updates?

@paulrostorp Could you please merge in master to your branch to fix CI & conflicts?

enahum commented:
@radex Sorry got caught up with some other issue that needs fixing in our product. I should be able to try this today

enahum commented:
Too many issues to list them in detail but here is a summary of what I'm getting

```
ound 452 errors in 277 files.

Errors  Files
     1  app/actions/local/channel.test.ts:392
     1  app/actions/local/channel.ts:241
     1  app/actions/local/team.ts:8
     1  app/actions/local/thread.ts:22
     1  app/actions/local/user.ts:13
     5  app/actions/remote/channel.ts:152
     1  app/actions/remote/post.ts:36
     1  app/actions/remote/reactions.ts:47
     1  app/actions/remote/retry.ts:193
     1  app/actions/remote/search.ts:15
     1  app/actions/remote/team.ts:81
     1  app/actions/remote/thread.ts:20
     2  app/actions/websocket/channel.ts:52
     1  app/actions/websocket/teams.ts:152
     1  app/components/announcement_banner/index.ts:4
     1  app/components/autocomplete/at_mention/index.ts:4
     1  app/components/autocomplete/channel_mention/index.ts:4
     1  app/components/autocomplete/channel_mention_item/index.ts:4
     1  app/components/autocomplete/emoji_suggestion/index.ts:4
     1  app/components/autocomplete/slash_suggestion/app_slash_suggestion/index.ts:4
     1  app/components/autocomplete/slash_suggestion/index.ts:4
     1  app/components/autocomplete_selector/index.tsx:4
     1  app/components/channel_actions/copy_channel_link_box/index.ts:7
     1  app/components/channel_actions/copy_channel_link_option/index.ts:4
     1  app/components/channel_actions/favorite_box/index.ts:7
     1  app/components/channel_actions/index.ts:4
     1  app/components/channel_actions/leave_channel_label/index.ts:7
     1  app/components/channel_actions/mute_box/index.ts:7
     2  app/components/channel_actions/set_header_box/index.ts:7
     1  app/components/channel_icon/dm_avatar/index.ts:4
     1  app/components/channel_item/custom_status/index.ts:4
     1  app/components/channel_item/index.ts:4
     1  app/components/channel_list_row/index.test.tsx:11
     1  app/components/common_post_options/copy_permalink_option/index.tsx:4
     1  app/components/common_post_options/follow_thread_option/index.ts:4
     1  app/components/custom_status/custom_status_emoji.test.tsx:11
     1  app/components/custom_status/custom_status_expiry.tsx:4
     2  app/components/emoji/emoji.tsx:4
     1  app/components/files/index.ts:4
     5  app/components/markdown/at_mention/at_mention.tsx:101
     1  app/components/markdown/at_mention/index.ts:4
     1  app/components/markdown/channel_mention/index.ts:4
     1  app/components/markdown/index.ts:4
     1  app/components/markdown/markdown_link/index.ts:4
     1  app/components/post_draft/draft_handler/index.ts:4
     1  app/components/post_draft/index.ts:4
     1  app/components/post_draft/post_input/index.ts:4
     1  app/components/post_draft/quick_actions/index.ts:4
     2  app/components/post_draft/send_handler/index.ts:4
     1  app/components/post_list/combined_user_activity/index.ts:4
     1  app/components/post_list/index.ts:4
     1  app/components/post_list/more_messages/index.ts:4
     1  app/components/post_list/post/avatar/index.ts:4
     1  app/components/post_list/post/body/add_members/index.ts:4
     1  app/components/post_list/post/body/content/embedded_bindings/button_binding/index.tsx:4
     1  app/components/post_list/post/body/content/embedded_bindings/menu_binding/index.tsx:4
     1  app/components/post_list/post/body/content/image_preview/index.ts:4
     1  app/components/post_list/post/body/content/opengraph/index.ts:4
     1  app/components/post_list/post/body/message/index.ts:4
     1  app/components/post_list/post/body/reactions/index.ts:4
     1  app/components/post_list/post/footer/index.ts:4
     1  app/components/post_list/post/header/index.ts:4
     1  app/components/post_list/post/index.ts:4
     1  app/components/post_list/post/system_message/index.tsx:4
     1  app/components/post_list/thread_overview/index.ts:4
     1  app/components/post_with_channel_info/channel_info/index.ts:4
     1  app/components/post_with_channel_info/index.ts:4
     1  app/components/server_version/index.tsx:4
     1  app/components/system_header/index.tsx:4
     1  app/components/team_list/team_list_item/index.ts:4
     1  app/components/team_sidebar/team_list/index.ts:6
     1  app/components/team_sidebar/team_list/team_item/index.ts:4
     1  app/components/threads_button/index.ts:4
     1  app/components/user_avatars_stack/user_avatar/index.ts:4
     1  app/components/user_item/index.ts:4
     1  app/components/user_list/index.test.tsx:11
     1  app/context/theme/index.tsx:14
     1  app/context/user_locale/index.tsx:13
     1  app/database/components/index.tsx:4
     4  app/database/manager/__mocks__/index.ts:5
    30  app/database/manager/index.ts:5
     1  app/database/migration/app/index.ts:4
     1  app/database/migration/server/index.ts:7
     1  app/database/models/app/global.ts:5
     1  app/database/models/app/info.ts:5
     1  app/database/models/app/servers.ts:5
    13  app/database/models/server/category.ts:5
     3  app/database/models/server/category_channel.ts:4
     5  app/database/models/server/channel.ts:4
     2  app/database/models/server/channel_info.ts:4
     5  app/database/models/server/channel_membership.ts:5
     1  app/database/models/server/config.ts:5
     1  app/database/models/server/custom_emoji.ts:5
     3  app/database/models/server/draft.ts:4
     4  app/database/models/server/file.ts:4
     9  app/database/models/server/group.ts:5
     3  app/database/models/server/group_channel.ts:4
     3  app/database/models/server/group_membership.ts:4
     3  app/database/models/server/group_team.ts:4
     4  app/database/models/server/my_channel.ts:4
     3  app/database/models/server/my_channel_settings.ts:4
     3  app/database/models/server/my_team.ts:4
    12  app/database/models/server/post.ts:5
     3  app/database/models/server/posts_in_channel.ts:4
     3  app/database/models/server/posts_in_thread.ts:4
     3  app/database/models/server/preference.ts:4
     3  app/database/models/server/reaction.ts:4
     1  app/database/models/server/role.ts:5
     1  app/database/models/server/system.ts:5
     7  app/database/models/server/team.ts:5
     3  app/database/models/server/team_channel_history.ts:4
     5  app/database/models/server/team_membership.ts:5
     3  app/database/models/server/team_search_history.ts:4
     3  app/database/models/server/team_threads_sync.ts:4
     3  app/database/models/server/thread.ts:4
     3  app/database/models/server/thread_in_team.ts:4
     3  app/database/models/server/thread_participant.ts:4
     5  app/database/models/server/user.ts:4
     1  app/database/operator/base_data_operator/index.ts:14
     2  app/database/operator/server_data_operator/handlers/channel.ts:174
     1  app/database/operator/server_data_operator/handlers/index.ts:51
     3  app/database/operator/server_data_operator/handlers/post.ts:18
     3  app/database/operator/server_data_operator/handlers/thread.ts:15
     1  app/database/operator/server_data_operator/transformers/channel.test.ts:112
     5  app/database/operator/server_data_operator/transformers/channel.ts:95
     1  app/database/operator/server_data_operator/transformers/general.test.ts:82
     5  app/database/operator/server_data_operator/transformers/general.ts:29
     1  app/database/operator/server_data_operator/transformers/index.ts:6
     1  app/database/operator/utils/test.ts:98
     1  app/database/operator/utils/thread.ts:7
     1  app/database/schema/server/table_schemas/config.ts:8
     1  app/database/schema/server/table_schemas/team_threads_sync.ts:8
     1  app/database/schema/server/table_schemas/thread.ts:8
     1  app/database/schema/server/table_schemas/thread_in_team.ts:8
     1  app/database/schema/server/table_schemas/thread_participant.ts:8
     2  app/helpers/database/index.ts:6
     1  app/products/calls/components/calls_custom_message/index.ts:4
     1  app/products/calls/components/channel_info_start/index.ts:4
     1  app/products/calls/components/join_call_banner/index.ts:4
     8  app/queries/servers/channel.ts:61
     2  app/queries/servers/custom_emoji.ts:11
     2  app/queries/servers/entry.ts:70
     1  app/queries/servers/team.ts:51
     1  app/queries/servers/thread.ts:18
     1  app/screens/browse_channels/index.ts:4
     1  app/screens/channel/channel_post_list/index.ts:5
     1  app/screens/channel/channel_post_list/intro/direct_channel/group/index.ts:4
     1  app/screens/channel/channel_post_list/intro/direct_channel/index.ts:4
     1  app/screens/channel/channel_post_list/intro/direct_channel/member/index.ts:4
     1  app/screens/channel/channel_post_list/intro/index.ts:4
     1  app/screens/channel/channel_post_list/intro/public_or_private_channel/index.ts:4
     2  app/screens/channel/header/index.ts:4
     1  app/screens/channel/index.tsx:4
     1  app/screens/channel_info/app_bindings/index.tsx:4
     1  app/screens/channel_info/destructive_options/archive/index.ts:4
     1  app/screens/channel_info/destructive_options/convert_private/index.ts:4
     2  app/screens/channel_info/extra/index.ts:4
     1  app/screens/channel_info/index.ts:4
     1  app/screens/channel_info/options/ignore_mentions/index.ts:4
     2  app/screens/channel_info/options/members/index.ts:4
     1  app/screens/channel_info/options/notification_preference/index.ts:4
     2  app/screens/channel_info/options/pinned_messages/index.ts:4
     1  app/screens/channel_info/title/direct_message/index.ts:4
     1  app/screens/channel_info/title/group_message/avatars/index.ts:4
     1  app/screens/channel_info/title/group_message/index.ts:4
     1  app/screens/channel_info/title/index.ts:4
     2  app/screens/channel_info/title/public_private/index.ts:4
     1  app/screens/create_direct_message/index.ts:4
     1  app/screens/create_or_edit_channel/index.tsx:4
     1  app/screens/custom_status/components/custom_status_suggestion/index.ts:4
     1  app/screens/custom_status/index.tsx:4
     1  app/screens/custom_status_clear_after/components/date_time_selector.tsx:4
     1  app/screens/custom_status_clear_after/index.tsx:4
     1  app/screens/edit_post/index.tsx:4
     1  app/screens/edit_profile/index.ts:4
     1  app/screens/emoji_picker/picker/filtered/index.ts:4
     1  app/screens/emoji_picker/picker/header/index.ts:4
     1  app/screens/emoji_picker/picker/index.ts:4
     1  app/screens/find_channels/filtered_list/index.ts:4
     1  app/screens/find_channels/filtered_list/remote_channel_item/index.ts:4
     1  app/screens/find_channels/filtered_list/user_item/index.ts:4
     1  app/screens/find_channels/quick_options/index.ts:4
     1  app/screens/find_channels/unfiltered_list/index.ts:4
     1  app/screens/gallery/document_renderer/index.ts:4
     1  app/screens/gallery/footer/index.ts:4
     1  app/screens/global_threads/index.tsx:4
     1  app/screens/global_threads/threads_list/index.ts:4
     1  app/screens/global_threads/threads_list/thread/index.ts:4
     1  app/screens/global_threads/threads_list/thread/thread_footer/index.ts:4
     1  app/screens/home/account/components/options/custom_status/custom_label/index.ts:4
     1  app/screens/home/account/index.ts:4
     1  app/screens/home/channel_list/additional_tablet_view/index.ts:4
     1  app/screens/home/channel_list/categories_list/categories/body/index.ts:5
     1  app/screens/home/channel_list/categories_list/categories/categories.test.tsx:11
     1  app/screens/home/channel_list/categories_list/categories/index.ts:4
     1  app/screens/home/channel_list/categories_list/categories/unreads/index.ts:4
     1  app/screens/home/channel_list/categories_list/header/index.ts:4
     1  app/screens/home/channel_list/categories_list/index.test.tsx:15
     1  app/screens/home/channel_list/categories_list/load_channels_error/index.ts:4
     1  app/screens/home/channel_list/categories_list/subheader/index.ts:4
     1  app/screens/home/channel_list/categories_list/subheader/unread_filter/index.ts:4
     1  app/screens/home/channel_list/index.ts:4
     2  app/screens/home/index.tsx:140
     1  app/screens/home/recent_mentions/index.ts:5
     1  app/screens/home/saved_messages/index.ts:5
     1  app/screens/home/search/index.tsx:4
     1  app/screens/home/search/initial/index.tsx:3
     1  app/screens/home/search/results/file_options/mobile_options.tsx:33
     1  app/screens/home/search/results/file_options/option_menus/index.tsx:4
     1  app/screens/home/search/results/file_options/tablet_options.tsx:79
     1  app/screens/home/search/results/index.tsx:4
     2  app/screens/home/search/search.tsx:246
     1  app/screens/home/tab_bar/account.tsx:4
     1  app/screens/integration_selector/channel_list_row/index.test.tsx:12
     1  app/screens/integration_selector/custom_list/index.test.tsx:13
     1  app/screens/integration_selector/custom_list_row/index.test.tsx:13
     1  app/screens/integration_selector/index.ts:3
     1  app/screens/integration_selector/option_list_row/index.test.tsx:12
     1  app/screens/integration_selector/selected_option/index.test.tsx:12
     1  app/screens/integration_selector/selected_options/index.test.tsx:12
     1  app/screens/invite/index.ts:4
     1  app/screens/join_team/index.ts:4
     1  app/screens/permalink/index.ts:4
     1  app/screens/pinned_messages/index.ts:4
     1  app/screens/post_options/index.ts:4
     1  app/screens/post_options/options/app_bindings_post_option.tsx:4
     1  app/screens/post_options/options/mark_unread_option/index.ts:4
     1  app/screens/post_options/reaction_bar/index.ts:4
     1  app/screens/reactions/index.ts:4
     1  app/screens/reactions/reactors_list/reactor/index.ts:4
     1  app/screens/select_team/index.ts:4
     1  app/screens/settings/about/index.tsx:4
     1  app/screens/settings/display/index.tsx:4
     1  app/screens/settings/display_clock/index.tsx:4
     1  app/screens/settings/display_crt/index.ts:4
     1  app/screens/settings/display_theme/index.ts:4
     1  app/screens/settings/display_timezone/index.tsx:4
     1  app/screens/settings/index.tsx:4
     1  app/screens/settings/notification_auto_responder/index.ts:4
     1  app/screens/settings/notification_email/index.tsx:4
     1  app/screens/settings/notification_mention/index.tsx:4
     1  app/screens/settings/notification_push/index.tsx:4
     1  app/screens/settings/notifications/index.tsx:4
     1  app/screens/settings/report_problem/index.ts:4
     1  app/screens/terms_of_service/index.ts:4
     1  app/screens/thread/index.tsx:4
     1  app/screens/thread/thread_follow_button/index.ts:4
     1  app/screens/thread/thread_post_list/index.ts:4
     1  app/screens/thread_options/index.ts:4
     1  app/screens/user_profile/index.ts:4
     3  app/utils/file/index.ts:4
     2  test/intl-test-helper.tsx:4
     2  types/database/models/servers/category.ts:4
     1  types/database/models/servers/category_channel.ts:8
     2  types/database/models/servers/channel.ts:14
     1  types/database/models/servers/channel_info.ts:6
     1  types/database/models/servers/channel_membership.ts:7
     1  types/database/models/servers/custom_emoji.ts:4
     1  types/database/models/servers/draft.ts:5
     1  types/database/models/servers/file.ts:6
     2  types/database/models/servers/group.ts:4
     1  types/database/models/servers/group_channel.ts:7
     1  types/database/models/servers/group_membership.ts:7
     1  types/database/models/servers/group_team.ts:7
     1  types/database/models/servers/post.ts:12
     1  types/database/models/servers/posts_in_channel.ts:6
     1  types/database/models/servers/posts_in_thread.ts:6
     1  types/database/models/servers/preference.ts:6
     1  types/database/models/servers/reaction.ts:7
     1  types/database/models/servers/team.ts:11
     1  types/database/models/servers/team_membership.ts:7
     1  types/database/models/servers/team_search_history.ts:6
     1  types/database/models/servers/team_threads_sync.ts:6
     1  types/database/models/servers/thread.ts:8
     1  types/database/models/servers/thread_in_team.ts:7
     1  types/database/models/servers/thread_participant.ts:7
     2  types/database/models/servers/user.ts:12
```

IMHO this are too many issues, so I think it would be best to revert the PR that got merged about the type definitions and perhaps test the new definitions against a project that already uses the previous ones.

If you want to test it against our project that is totally fine :) https://github.com/mattermost/mattermost-mobile ;)

radex commented:
@paulrostorp How do you want to proceed with this PR?

paulrostorp commented:
@enahum sorry for the delay, I was off a couple of days. How are you producing this list ? Are you sure you are using the right typescript executable ? Are you sure you have installed the code in this branch properly ?

@radex I am quite confident in the code in PR, it works both on real projects but also via tests on my machine and in CI. 

I'll try testing against mattermost's repo to find out if anything unexpected is going on. I'll ping you later about merging this

paulrostorp commented:
@enahum I'm a bit confused as to your setup...
How are you producing that list of errors ? If using VSCode, try reloading it, that makes most errors disappear after installing the custom version of watermelonDB

I was able to reproduce only 3 errors in your repo (affecting 230 files only, not 277): 

1. Most of 99% of those where because of a single error: a bug with`withDatabase`: the type was too strict, I've relaxed it now and it resolved all those errors. Addressed in https://github.com/Nozbe/WatermelonDB/pull/1477/commits/7f629e6b5b9c9b5ace7800110f11c77c64eb857f
2. The `Database` constructor's `modelClasses`: following [your comment](https://github.com/Nozbe/WatermelonDB/pull/1477#discussion_r1084363501), I had reverted the type to accept `Class<Model>[]` but it seems your project doesn't like that... I've made the type more flexible to accept both `Class<Model>[]` (as it was on v0.24) and `Model[]` (which your project requires). Addressed in https://github.com/Nozbe/WatermelonDB/pull/1477/commits/90b8997c88b0db74cb776bdf76e5fd18c272d194

3. An error in `app/actions/local/channel.test.ts`, which is justified, your code calls a `model.cancelPrepareUpdate()`, that method doesn't exist...


@radex Following these two commits, I think we can merge this PR

radex commented:
@paulrostorp just a request, please don't force-push here, it makes it difficult to review your new commits

radex commented:
I think what I'll do is use @enahum's minimal patch for 0.25.2 to fix the most glaring 0.25 TS issues, and merge this PR to master to give people more time to test pre-releases and find any remaining issues.

It does concern me very much that with many TS pull requests, what fixes it for some users breaks it for others ;)

enahum commented:
I'll re-check with the latest changes soon and yes the function that does not exist is being added by us :)

paulrostorp commented:
@radex Sorry about the force push, I was just rebasing to fix the conflicts.

1. The patch provided by @enahum fixed only a small set of issues brought on by https://github.com/Nozbe/WatermelonDB/pull/1318 in our codebase.
2. Most of what my PR does is revert PR https://github.com/Nozbe/WatermelonDB/pull/1318 but also:
  1. Incorporate types added by https://github.com/Nozbe/WatermelonDB/pull/1318 for things missing types
  2. Add testing to protect against future regressions of types.

3. I think that with the tests added in this PR regressions should become less frequent. But I would advocate for a typescript rewrite in this repo if the burden of maintaining separate types becomes too heavy; typings is rarely a concern in typescript-first projects, but that is a question of opinion I guess

radex commented:
@paulrostorp Thanks for explanation, your work will certainly not go to waste, the question is merely whether to release it as part of a patch release (fixing some issues, risking introducing new ones) or as part of a minor release.

Typescript rewrite is _currently_ out of the question. I'm not ruling it out for the future, but currently, both TS and Flow need to be supported, and it's simply easier for me to keep Flow as the primary typing system and TS as the community-maintained alternative typings set.

enahum commented:
Ok I just checked the latest commit and is working just fine @paulrostorp 
cc: @radex 

Thanks guys

radex commented:
@paulrostorp If you're interested in contributing more, there's a few more TS-related maintenance items:

- prettier doesn't like the current .d.ts files
- in package.json there are outdated TS-related dependencies… I'm not even sure that all of them should be there, given that much of the TS checking lives in examples/typescript

mlecoq commented:
@paulrostorp I have one issue left 

```export type DirtyRaw = { [id: string]: any }```

has been changed to 

```export type DirtyRaw = Object```

causing issues in synchronisation code, did you notice this one ?

mlecoq commented:
Another issue is on `Model`

```
(property) Model.id: any
Cannot assign to 'id' because it is a read-only property
```

and I need to be able to set it manually in some cases

radex commented:
@mlecoq regarding `Model.id` - the error is correct. `id` does not have a setter. Unless you overrode it yourself, you can't set it manually _this_ way

mlecoq commented:
> @mlecoq regarding `Model.id` - the error is correct. `id` does not have a setter. Unless you overrode it yourself, you can't set it manually _this_ way

yes, older types was misleading about ids, we thought it worked but in fact given id in our code was ignored

paulrostorp commented:
@mlecoq Regarding the `DirtyRaw` type, the `Object` type is more relaxed than `{ [id: string]: any }` so it shouldn't make use more difficult. Could you share what error you receive in sync ?

mlecoq commented:
@paulrostorp in `pushChanges` method I need to make some transformations to adapt stored data to the graphql API

for each accessed property in dirty raw, I have the following ts error 

```Property '<nameOfProperty>' does not exist on type 'Object'```

paulrostorp commented:
@mlecoq Ah I see, I thought you were referring to the `pull` part of sync.
I personally recommend you do some type checking, maybe make use of [type guards](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) or if you are confident in the type, you can cast it to whatever you need. the `Record<string, any>` is not very safe

For example:
```ts
const isProject = (dirty: Object): dirty is Project => {
  return "id" in dirty && typeof dirty.id == "string" && "title" in dirty;
};
```

But I do believe the type is guaranteed to have an `id`, so maybe we could define that in the type... @radex please advise

radex commented:
DirtyRaw is more of an human-readable annotation than a real type. The only thing it guarantees is that it's an object (i.e. not null, and not something else, and you can do `x[y]` or `...x` without crashing) - it's an unsanitized (dirty, user-provided) input to functions that return a `RawRecord`, i.e. an object that has a guaranteed structure

paulrostorp commented:
> The only thing it guarantees is that it's an object (i.e. not null, and not something else

In that case I think the proper type is `object`. 

I am now realizing https://github.com/Nozbe/WatermelonDB/pull/1318 set the value to `Object` (the class instance) instead of lowercase `object`

`object` is the proper type, I'll make a PR

paulrostorp commented:
@radex https://github.com/Nozbe/WatermelonDB/pull/1509

radex commented:
ah, that makes sense 👍. In Flow, it _is_ "Object" (kinda, sorta)

radex commented:
contd. #1516
